### PR TITLE
Small changes for improving training

### PR DESCRIPTION
**Pull request by josauder**
- Add linear learning rate warmup (https://arxiv.org/abs/1706.02677) defaults to 2000 iterations
- Add mixed precision training (amp bfloat16), which mixes 16 and 32 bit floats depending on the numerics and allows training at larger batch sizes.
- Make .detach() on the concept logits configurable, which stops gradient flow from the classification layer through the concept layer: there was an unwanted interaction with the `not_given` multi-hot concept.

### CODE REVIEW

**Review by yeelauren — APPROVED** [review]
LGTM -- were you able to test and get numbers for the speed up ? Just curious@

**Comment by josauder** [comment]
Almost all of the forward parts (data loading, forward passes) are basically perfectly parallelizable, and we can run (the forward pass of) almost any batch size in ~1-1.3s on an A100 GPU. The difference is in the batch size that we can fit into the GPU. So iterating through a dataset of 100k images with batch size 20 will be 2x faster than iterating through the same dataset with batch size 10 - using mixed precision training allows you to double the batch size. 

For the other changes:
1. LR warmup is used in basically every big model these days, and allows for more stable training with larger learning rates (better generalization with shorter training times)
2. The `.detach()` is not about speed, but the problem is that because we have multi-hot concepts with `not_given`, backpropagating through these concepts meant that the model was using these concepts to "cheat" for the classification.


The easiest way to further speed up training (increase the batch size), is reducing the resolution.Right now, the GPU VRAM breaks down into the following:

1. DINOv3 encoder weights (0.3GB for ViT-B or 1.2GB for ViT-L) - we don't compute a gradient on the encoder so this is **resolution & batch_size independent**
2. LoRA weights+gradients for the DINOv3 encoder roughly `0.02 * batch_size * size_of_encoder_weights`. This means we train the equivalent of approximately 2% of the weights of the encoder, so it is also a negligible size, even when using ViT-L. This is still **resolution independent**
4. DPT head weights+gradients + concept->class layer. This part is **resolution and batch_size dependent**, as the DPT head+gradients scale from the output tokens (H//16xW//16) to the output resolution HxW, and so we have to store extremely large gradients tensors e.g. `batch_size x 512 x 512 x num_concepts x num_classes` for the last layer. They are very fast and cheap to compute in the forward pass, but the gradient calculation takes a lot of memory. 

Number 3 is by far the dominating factor of the VRAM: Halving the training resolution allows you to quadruple the batch size.
